x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java

```diff
@@ -0,0 +1,339 @@
+
+    public static class Builder extends FieldMapper.Builder<AggregateMetricFieldMapper.Builder, AggregateMetricFieldMapper> {
+
+        private Boolean ignoreMalformed;
+
+        private Set<String> metrics;
+
+        public Builder(String name) {
+            super(name, Defaults.FIELD_TYPE, Defaults.FIELD_TYPE);
+            builder = this;
+        }
+
+        public AggregateMetricFieldMapper.Builder ignoreMalformed(boolean ignoreMalformed) {
+            this.ignoreMalformed = ignoreMalformed;
+            return builder;
+        }
+
+        protected Explicit<Boolean> ignoreMalformed(BuilderContext context) {
+            if (ignoreMalformed != null) {
+                return new Explicit<>(ignoreMalformed, true);
+            }
+            if (context.indexSettings() != null) {
+                return new Explicit<>(IGNORE_MALFORMED_SETTING.get(context.indexSettings()), false);
+            }
+            return AggregateMetricFieldMapper.Defaults.IGNORE_MALFORMED;
+        }
+
+        public AggregateMetricFieldMapper.Builder metrics(Set<String> metrics) {
+            this.metrics = metrics;
+            return builder;
+        }
+
+        protected Explicit<Set<String>> metrics(BuilderContext context) {
+            if (metrics != null) {
+                return new Explicit<>(metrics, true);
+            }
+            return Defaults.METRICS;
+        }
+
+        @Override
+        public AggregateMetricFieldMapper build(BuilderContext context) {
+            setupFieldType(context);
+            return new AggregateMetricFieldMapper(name, fieldType, defaultFieldType,
+                context.indexSettings(), multiFieldsBuilder.build(this, context),
+                ignoreMalformed(context), metrics(context), copyTo);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        @Override
+        public Mapper.Builder<Builder, AggregateMetricFieldMapper> parse(String name,
+                                                                         Map<String, Object> node,
+                                                                         ParserContext parserContext) throws MapperParsingException {
+            AggregateMetricFieldMapper.Builder builder = new AggregateMetricFieldMapper.Builder(name);
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext(); ) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String propName = entry.getKey();
+                Object propNode = entry.getValue();
+                if (propName.equals(Names.METRICS.getPreferredName())) {
+                    String metricsStr[] = XContentMapValues.nodeStringArrayValue(propNode);
+                    // Convert the array of Metric enum to a Set of their string representation
+                    Set<String> supporteddMetrics = Arrays.stream(Metrics.values()).map(s -> s.toString()).collect(Collectors.toSet());
```

Comment:
I wonder if we should use an `EnumSet` here rather than converting back to strings?  Or implementing something like [`fromString()` on the enum](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java#L68) to parse the string and return the enum (otherwise throw an exception).

Since we already have the enum it'd be nice if we could convert into enum space rather than other way around.

Reply:
Fixed